accesshandler/pkg/api/grants.go

```diff
@@ -19,31 +21,40 @@ func (a *API) PostGrants(w http.ResponseWriter, r *http.Request) {
 	ctx := r.Context()
 	var b types.CreateGrant
 
-	grant, err := func() (*types.Grant, error) {
-		err := apio.DecodeJSONBody(w, r, &b)
+	err := apio.DecodeJSONBody(w, r, &b)
+	if err != nil {
+		apio.Error(ctx, w, err)
+		return
+	}
+	_, ok := config.Providers[b.Provider]
+	if !ok {
+		err = apio.NewRequestError(errors.New("provider does not exist"), http.StatusBadRequest)
 		if err != nil {
-			return nil, err
+			apio.Error(ctx, w, err)
+			return
 		}
-
-		g, err := b.Validate(ctx, a.Clock.Now())
+		return
+	}
+	g, err := b.Validate(ctx, a.Clock.Now())
+	if err != nil {
+		// return the error details to the client if validation failed
+		err = apio.NewRequestError(err, http.StatusBadRequest)
 		if err != nil {
-			// return the error details to the client if validation failed
-			return nil, &apio.APIError{
-				Err:    err,
-				Status: http.StatusBadRequest,
-			}
+			apio.Error(ctx, w, err)
+			return
 		}
+		return
+	}
 
-		return a.runtime.CreateGrant(ctx, *g)
-	}()
-
+	grant, additionalProperties, err := a.runtime.CreateGrant(ctx, *g)
```

Comment:
I understand the approach you're taking here but I'm not convinced that adding `additionalProperties` to our Access Provider framework is a good idea.

It seems that `additionalProperties` is used as a generic output mechanism where the Access Provider has generated some data which we wish to store. Currently our only use case for `additionalProperties` is allowing the flask provider to output an Access Token.

However, Access Tokens are not just any old data - they have special meaning and need to be stored in our database so that our webhook handler can look it up. Using `additionalProperties` for this means that we're introducing a generic framework feature which can be confusing (as an Access Provider developer, the question I'd ask is - when should I be using `additionalProperties`), which at the same time has specific meaning.

This approach has led to the Access Token implementation being spread across 3 parts of the codebase:
- the Approvals API (handling the Access Token in the `additionalProperties` response)
- the Access Handler API (generating the Access Token)
- the provider itself (implementing `RequiresAccessToken`)

On the [YAGNI principle](https://martinfowler.com/bliki/Yagni.html) I don't think we should take this approach. My main opposition is that we're introducing a generic "additionalProperties" field without defined use cases on how this will be used across many Access Providers.

Here's my proposed approach which I think makes our framework simpler:

- when creating a request, the Approvals API checks whether the provider implements `RequiresAccessToken`. If it does, the Approvals API generates the Access Token and saves it in the database.

This requires the Approvals API to know about which providers are installed - but it already does! The API now has access to the provider metadata and can look up a provider without needing to fully initialise it. You can have a look at the `providersetup` endpoints I wrote which utilise this.

Separately, I do think the idea of having storage associated with a Grant can be useful, and I think we'll return to this to design this out in future once we have some concrete use cases in mind. I don't think Access Tokens are part of this generic storage though.